services/rfq/relayer/pricer/fee_pricer.go

```diff
@@ -147,11 +157,95 @@ func (f *feePricer) GetDestinationFee(parentCtx context.Context, _, destination
 		fee = new(big.Int).Add(fee, l1Fee)
 		span.SetAttributes(attribute.String("l1_fee", l1Fee.String()))
 	}
+
+	// If specified, calculate and add the call fee, as well as the call value which will be paid by the relayer
+	if tx != nil {
+		fee, err = f.addZapFees(ctx, destination, denomToken, isQuote, tx, fee)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	span.SetAttributes(attribute.String("destination_fee", fee.String()))
 	return fee, nil
 }
 
-func (f *feePricer) GetTotalFee(parentCtx context.Context, origin, destination uint32, denomToken string, isQuote bool) (_ *big.Int, err error) {
+//nolint:gosec
+func (f *feePricer) addZapFees(ctx context.Context, destination uint32, denomToken string, isQuote bool, tx *fastbridgev2.IFastBridgeV2BridgeTransactionV2, fee *big.Int) (*big.Int, error) {
+	span := trace.SpanFromContext(ctx)
+
+	if len(tx.CallParams) > 0 {
+		gasEstimate, err := f.getZapGasEstimate(ctx, destination, tx)
+		if err != nil {
+			return nil, err
+		}
+		callFee, err := f.getFee(ctx, destination, destination, int(gasEstimate), denomToken, isQuote)
+		if err != nil {
+			return nil, err
+		}
+		fee = new(big.Int).Add(fee, callFee)
+		span.SetAttributes(attribute.String("call_fee", callFee.String()))
+	}
+
+	if tx.CallValue != nil && tx.CallValue.Sign() > 0 {
+		callValueFloat := new(big.Float).SetInt(tx.CallValue)
+		valueDenom, err := f.getDenomFee(ctx, destination, destination, denomToken, callValueFloat)
+		if err != nil {
+			return nil, err
+		}
+		valueScaled, err := f.getFeeWithMultiplier(ctx, destination, isQuote, valueDenom)
+		if err != nil {
+			return nil, err
+		}
+		fee = new(big.Int).Add(fee, valueScaled)
+		span.SetAttributes(attribute.String("value_scaled", valueScaled.String()))
+	}
+
+	return fee, nil
+}
+
+// cache so that we don't have to parse the ABI every time.
+var recipientABI *abi.ABI
+
+const methodName = "fastBridgeTransferReceived"
+
+func (f *feePricer) getZapGasEstimate(ctx context.Context, destination uint32, tx *fastbridgev2.IFastBridgeV2BridgeTransactionV2) (gasEstimate uint64, err error) {
+	client, err := f.clientFetcher.GetClient(ctx, big.NewInt(int64(destination)))
+	if err != nil {
+		return 0, fmt.Errorf("could not get client: %w", err)
+	}
+
+	if recipientABI == nil {
+		parsedABI, err := abi.JSON(strings.NewReader(fastbridgev2.IFastBridgeRecipientMetaData.ABI))
+		if err != nil {
+			return 0, fmt.Errorf("could not parse ABI: %w", err)
+		}
+		recipientABI = &parsedABI
+	}
+	destAmount := tx.DestAmount
+	if destAmount == nil {
+		destAmount = big.NewInt(0)
+	}
+	encodedData, err := recipientABI.Pack(methodName, tx.DestToken, destAmount, tx.CallParams)
+	if err != nil {
+		return 0, fmt.Errorf("could not encode function call: %w", err)
+	}
+
+	callMsg := ethereum.CallMsg{
+		From:  f.relayerAddress,
+		To:    &tx.DestRecipient,
+		Value: tx.CallValue,
+		Data:  encodedData,
+	}
+	gasEstimate, err = client.EstimateGas(ctx, callMsg)
```

Comment:
We should be estimating the gas cost for the whole `relay()` call here instead.

Estimating gas for the recipient call should _probably_ work for the native gas token (assuming the recipient doesn't implement any caller checks, e.g. only allows the FastBridge to call the hook function), but will definitely fail for ERC20s. These are transferred separately by the `FastBridge` contract (as the ERC20 standard doesn't have built-in "transfer and call"), so just calling the hook function will result in a revert (as the recipient doesn't yet have the tokens transferred to them), leading to infinite gas estimation (or an error - not sure about the exact behaviour there).